syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.InventoryRecord";
option java_outer_classname = "InventoryRecordProto";
option objc_class_prefix = "RTG";


service InventoryRecordService {
  //Create a rpc method to get the inventory record by searching ID
  rpc getInventoryRecordById (InventoryRecordRequest) returns (InventoryRecordResponse) {}
  //Create a rpc method to get the inventory record by searching key value pair
  rpc getInventoryRecordByKeyValue (InventoryRecordRequestSearch) returns (InventoryRecordResponse) {}
  //Create a rpc method to get the inventory record by searching key value pair for range
  rpc getInventoryRecordByKeyValueRange (InventoryRecordSearchRange) returns (InventoryRecordResponseList) {}
  //Create a rpc method to get distribution in double for a key name and percentile
  rpc getDistribution (GetDistribution) returns (InventoryAtPercentile) {}
  //Create a rpc method to update a record an return boolean
  rpc updateInventoryRecord (UpdateInventoryRecord) returns (InventoryRecordResponse) {}
}

//Define the request message
message InventoryRecordRequest {
  string id = 1;
}

//Define key value pair request message
message InventoryRecordRequestSearch {
  string key_name = 1;
  string key_value = 2;
}

//Define key value pair request message for range
message InventoryRecordSearchRange {
  string key_name = 1;
  string key_value_min = 2;
  string key_value_max = 3;
}

//Define key value pair request message for distribution
message GetDistribution {
  string key_name = 1;
  double percentile = 2;
}

message UpdateInventoryRecord {
  InventoryRecord inventoryRecord = 1;
}

//Define the response message
message InventoryRecordResponse {
   InventoryRecord inventoryRecord = 1;
}

//Define the response message
message  InventoryRecordBoolResponse {
  bool result = 1;
}

//Define the response message list
message InventoryRecordResponseList {
  repeated InventoryRecord inventoryRecord = 1;
}

//Define the double response message
message InventoryAtPercentile {
  double at_percentile = 1;
}


//Define the InventoryRecord message
message InventoryRecord {
  string id = 1;
  string name = 2;
  string description = 3;
  float unitPrice = 4;
  int32 quantityInStock = 5;
  float inveneotryValue = 6;
  int32 reorderLevel = 7;
  int32 reorderTimeInDays = 8;
  int32 quantityInReorder = 9;
  bool discontinued = 10;
}